Fix/reduce Cppcheck warnings in test code

KaganCanSit opened the pull request:
Hello,

I reviewed the warnings reported by cppcheck under the test folder, which I think are not false positives. I tried to make corrections to the appropriate ones. The messages I reviewed and intervened are listed below:

* [botan/src/tests/test_pad.cpp:33] (performance) Ineffective call of function 'substr' because a prefix of the string is assigned to itself. Use resize() or pop_back() instead. [uselessCallsSubstr]
* [botan/src/tests/test_ec_group.cpp:115] (style) The scope of the variable 'blind_ws' can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. When you see this message it is always safe to reduce the variable scope to 1 level. [variableScope]
* [botan/src/tests/test_hash.cpp:196] (style) Variable 'buf' is assigned a value that is never used. [unreadVariable] - !!! I think this change and function needs to be examined.
* [botan/src/tests/test_pkcs11_high_level.cpp:1067] (style) Variable 'curve' can be declared as reference to const [constVariable] 
[botan/src/tests/test_pkcs11_high_level.cpp:1092] (style) Variable 'curve' can be declared as reference to const [constVariable]
* [botan/src/tests/test_pkcs11_low_level.cpp:389] (style) Local variable 'label' shadows outer variable [shadowVariable]
* [botan/src/tests/test_tls_rfc8448.cpp:253] (style) Consider using std::find_if algorithm instead of a raw loop. [useStlAlgorithm]
* [botan/src/tests/test_uri.cpp:48] (style) Variable 'tests' can be declared as const array [constVariable]
* [botan/src/tests/test_utils.cpp:1012] (style) Consider using std::transform algorithm instead of a raw loop. [useStlAlgorithm]
* [botan/src/tests/test_utils.cpp:1329] (style) Consider using std::fill algorithm instead of a raw loop. [useStlAlgorithm]

I did not make any changes for the following warnings:
- 'explicit' and 'is not initialized in the constructor' messages. (I can try to create a separate PR request by reviewing this if desired.)
- Situations where I think there will be no significant benefit in using the algorithm in various raw loops.

Compile:
```bash
python3 src/scripts/dev_tools/run_clang_format.py --clang-format=clang-format-17 --src-dir=src && ninja clean && ./configure.py --without-documentation --with-boost --cc=clang --compiler-cache=ccache --build-targets=static,cli,tests --build-tool=ninja && ninja
```

Test:
```bash
./botan-test --test-threads=4 --run-long-tests && python3 src/scripts/test_cli.py ./botan cli_tls_socket_tests
```

You can contact me for the points you want to change or improve. I would like to help.

Have a nice day.

reneme commented on src/tests/test_utils.cpp:
```diff
@@ -1006,11 +1006,9 @@ class Date_Format_Tests final : public Text_Based_Test {
-         for(const auto& sub : parts) {
-            u32s.push_back(Botan::to_u32bit(sub));
-         }
+         std::vector<uint32_t> u32s(parts.size());
+         std::transform(
+            parts.begin(), parts.end(), u32s.begin(), [](const std::string& sub) { return Botan::to_u32bit(sub); });
          return u32s;
       }
```
Such use of STL algorithms (especially pre-ranges) are really debatable in my opinion. Yes, using `std::transform`, `std::find_if`, `std::for_each`, and friends is more idiomatic. And there are plenty C++ connoisseurs that want to eliminate raw loops as much as possible. Though, I think we should ask ourselves: is that really easier to read than a range-based for loop?

For the record: I'm essentially indifferent about this. But I think its not worth putting in the effort to migrate to these pre-ranges STL algorithms. Once we shed support for some older compilers and can actually use the full potential of ranges, that's an entirely different discussion, in my opinion.

This would be C++23 ([godbolt](https://godbolt.org/z/a5xTEa7xj)):

```C++
namespace views = std::ranges::views;
namespace ranges = std::ranges;

auto u32s = parts | views::transform([](auto&& v) { return Botan::to_u32bit(v); })
                  | ranges::to<std::vector>();
```

... now, that would be real improvement! No more iterators, no more manual pre-allocation of the vector, not even explicit value types anymore. 🤩

KaganCanSit replied:
Frankly, there were similar warnings to this situation. Sometimes I think they made me think about readability. That's why I only took the ones that exactly match the call

The usage you gave as an example really excited me. I'll look into it. But as far as I understand, C++20 doesn't support this usage.

reneme replied:
> But as far as I understand, C++20 doesn't support this usage.

C++20 has ranges (just not all compilers we currently support). Chances are that Botan 4 will be able to use much of this. The only proper C++23-ism in my example above is `ranges::to<>`. But there's (afaik) no technical reason why it couldn't be custom built in C++20 until we adopt C++23 in the future.

KaganCanSit replied:
> C++20 has ranges (just not all compilers we currently support). Chances are that Botan 4 will be able to use much of this. The only proper C++23-ism in my example above is `ranges::to<>`. But there's (afaik) no technical reason why it couldn't be custom built in C++20 until we adopt C++23 in the future.

Thanks for sharing your knowledge with me. I will try this approach in a small test code.

reneme commented on src/tests/test_tls_rfc8448.cpp:
```diff
@@ -249,10 +249,10 @@ class Test_TLS_13_Callbacks : public Botan::TLS::Callbacks {
-            if(mock.message_to_sign == msg) {
-               return mock.signature_to_produce;
-            }
+         const auto found =
+            std::ranges::find_if(m_mock_signatures, [&msg](const auto& mock) { return mock.message_to_sign == msg; });
+         if(found != m_mock_signatures.end()) {
+            return found->signature_to_produce;
          }
```
See my other comment about pre-ranges STL algorithms.

KaganCanSit commented on src/tests/test_uri.cpp:
```diff
@@ -45,7 +45,9 @@ class URI_Tests final : public Test {
                Botan::URI::Type type;
                uint16_t port;
-         } tests[]{
+         }
+
+         const tests[]{
```
I understand, I wanted to implement it array based at first, but I thought it might not be welcomed because there would be too many changes.

I will give it a try after this message.

Thanks for the quick turnaround.

KaganCanSit commented:
@randombit @reneme Thank you very much for your quick feedback. I apologize for confusing the commit name while managing multiple branches. I corrected the PR and comment name.

* Rebased to master branch.
* STL algorithm calls were retrieved as requested.
* Renamed the struct structure and used std::array.
* added a comment line for code readability at the point where resize is selected for performance instead of substr.

I hope there is no point I missed. I will follow CI.

randombit reviewed (approved):
LGTM thanks
